### PR TITLE
fix(storage): fix compactor oom

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- https://github.com/risingwavelabs/risingwave/issues/11142

This pr fixes two major issues found in 11142 that caused oom
1. recv buffer stacking
2. XorFilter using a lot of memory during the build process.

In order to fix the above issues, we have introduced a number of changes
1. new memory estimation algorithm and memory request for input and ouput before the compact task starts.
2. introduce a new configuration to limit the size of the recv buffer.
3. introduce a new memory estimation algorithm for XorFilter and add a memory limit at finish time.
4. configure the parameters related to compact and s3

There are some tradeoffs. 
After testing, I found that adjusting the recv buffer doesn't cause any significant performance degradation, but it does reduce the memory footprint, so I've made it a default.

nexmark q16

- recv buffer default

image

- limit recv buffer
image

When we limit the recv buffer size, we can observe that by default the linked recv-q will not exceed 400k.
image

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/config.rs [hunk 1]
```diff
@@ -857,6 +888,10 @@ pub mod default {
         pub fn object_store_read_timeout_ms() -> u64 {
             60 * 60 * 1000
         }
+
+        pub fn compactor_max_sst_key_count() -> u64 {
+            200 * 10000 // 200w
```
Comment (Little-Wallace) [inline src/common/src/config.rs:893]: Prefer `2 * 1024 * 1024`

File: src/common/src/config.rs [hunk 2]
```diff
@@ -451,10 +451,15 @@ pub struct StorageConfig {
     #[serde(default = "default::storage::share_buffer_upload_concurrency")]
     pub share_buffer_upload_concurrency: usize,
 
-    /// Capacity of sstable meta cache.
     #[serde(default)]
     pub compactor_memory_limit_mb: Option<usize>,
 
+    #[serde(default = "default::storage::compactor_max_task_multiplier")]
+    pub compactor_max_task_multiplier: f32,
```
Comment (zwang28) [inline src/common/src/config.rs:458]: Please add comments for these 2 new config.

File: src/common/src/config.rs [hunk 3]
```diff
@@ -485,6 +490,24 @@ pub struct StorageConfig {
     #[serde(default = "default::storage::object_store_read_timeout_ms")]
     pub object_store_read_timeout_ms: u64,
 
+    #[serde(default = "default::s3_objstore_config::object_store_keepalive_ms")]
+    pub object_store_keepalive_ms: u64,
```
Comment (zwang28) [inline src/common/src/config.rs:494]: `object_store_keepalive_ms` can be an Option too.

File: src/common/src/config.rs [hunk 4]
```diff
@@ -821,6 +844,14 @@ pub mod default {
             512
         }
 
+        pub fn compactor_max_task_multiplier() -> f32 {
+            1.501
```
Comment (zwang28) [inline src/common/src/config.rs:848]: What's the 0.001 for?

File: src/storage/compactor/src/server.rs [hunk 5]
```diff
@@ -203,7 +204,7 @@ pub async fn compactor_serve(
             opts.compaction_worker_threads_number,
         )),
         filter_key_extractor_manager: filter_key_extractor_manager.clone(),
-        output_memory_limiter,
+        output_memory_limiter: memory_limiter,
```
Comment (zwang28) [inline src/storage/compactor/src/server.rs:207]: Should output_memory_limiter be renamed?

File: src/storage/src/hummock/compactor/mod.rs [hunk 6]
```diff
@@ -281,27 +282,36 @@ impl Compactor {
...
-                task_memory_capacity_with_parallelism
             );
 
         // If the task does not have enough memory, it should cancel the task and let the meta
         // reschedule it, so that it does not occupy the compactor's resources.
         let memory_detector = context
             .output_memory_limiter
-            .try_require_memory(task_memory_capacity_with_parallelism as u64);
+            .try_require_memory(task_memory_capacity_with_parallelism);
         if memory_detector.is_none() {
```
Comment (zwang28) [inline src/storage/src/hummock/compactor/mod.rs:315]: Could you explain why not `require_memory`?

File: src/storage/src/hummock/sstable/bloom.rs [hunk 7]
```diff
@@ -150,7 +152,7 @@ impl FilterBuilder for BloomFilterBuilder {
         self.key_hash_entries.len() * 4
     }
 
-    fn finish(&mut self) -> Vec<u8> {
+    fn finish(&mut self, _memory_limiter: Option<Arc<MemoryLimiter>>) -> Vec<u8> {
```
Comment (zwang28) [inline src/storage/src/hummock/sstable/bloom.rs:155]: Why not require memory for bloom filter?

File: src/storage/src/hummock/sstable_store.rs [hunk 8]
```diff
@@ -780,16 +780,20 @@ impl SstableWriter for StreamingUploadWriter {
                     t
                 });
 
+            debug_assert!(
+                !meta.block_metas.is_empty() || !meta.monotonic_tombstone_events.is_empty()
```
Comment (zwang28) [inline src/storage/src/hummock/sstable_store.rs:784]: What about assert!, or at least warn!. So that we're aware of its happening.

**Review by zwang28 — APPROVED** [review]
LGTM. Thanks.

**Review by Little-Wallace — APPROVED** [review]
LGTM

File: docker/Dockerfile [hunk 9]
```diff
@@ -42,9 +42,9 @@ RUN cargo fetch && \
   chmod +x /risingwave/bin/jeprof && \
   mkdir -p /risingwave/lib && cargo clean
 
-RUN cd /risingwave/java && mvn -B package -Dmaven.test.skip=true -Djava.binding.release=true && \
-    mkdir -p /risingwave/bin/connector-node && \
-    tar -zxvf /risingwave/java/connector-node/assembly/target/risingwave-connector-1.0.0.tar.gz -C /risingwave/bin/connector-node
+# RUN cd /risingwave/java && mvn -B package -Dmaven.test.skip=true -Djava.binding.release=true && \
+#     mkdir -p /risingwave/bin/connector-node && \
```
Comment (wcy-fdu) [inline docker/Dockerfile:46]: Why do we change dockerfile?

File: src/common/src/config.rs (hunk 4 above)
Comment (wcy-fdu) [reply src/common/src/config.rs:848]: +1

**Review by wcy-fdu — APPROVED** [review]
Rest LGTM!

File: docker/Dockerfile (hunk 9 above)
Comment (Li0k) [reply docker/Dockerfile:46]: i will revert it

**Pull request merged**